lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +170,65 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(table, columns, name_suffix: nil, if_not_exists: false)
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    schema, table = _schema_and_table_for(table)
+
+    is_partitioned = select_value(<<~SQL)
+      SELECT EXISTS (
+        SELECT 1
+        FROM pg_partitioned_table, pg_class, pg_namespace
+        WHERE pg_class.oid = pg_partitioned_table.partrelid
+          AND pg_class.relnamespace = pg_namespace.oid
+          AND pg_class.relname = #{quote(table)}
+          AND pg_namespace.nspname = #{quote(schema)}
+      )
+    SQL
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{table} is not a partitioned table" unless is_partitioned
+
+    name_suffix = "#{Array.wrap(columns).join("_")}_idx" unless name_suffix.present?
+
+    child_tables = select_values(<<~SQL)
+      SELECT child.relname
+      FROM pg_inherits
+        JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child  ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace    ON parent.relnamespace = pg_namespace.oid
+      WHERE parent.relname = #{quote(table)}
+        AND pg_namespace.nspname = #{quote(schema)}
+    SQL
+
+    parent_index = "#{table}_#{name_suffix}"
+
+    quoted_columns = Array.wrap(columns).map { |c| quote_column_name(c) }.join(", ")
+
+    unsafe_execute(<<~SQL)
+      CREATE INDEX#{if_not_exists ? " IF NOT EXISTS" : ""} #{quote_column_name(parent_index)}
+      ON ONLY #{quote_schema_name(schema)}.#{quote_table_name(table)} (#{quoted_columns})
+    SQL
+
+    child_tables.each do |child_table|
+      child_index = "#{child_table}_#{name_suffix}"
+
+      next if if_not_exists && _index_exists?(schema, child_index)
+
+      safe_add_concurrent_index("#{quote_schema_name(schema)}.#{child_table}", columns, name: child_index)
+
+      unsafe_execute(<<~SQL)
+        ALTER INDEX #{quote_schema_name(schema)}.#{quote_column_name(parent_index)}
+        ATTACH PARTITION #{quote_schema_name(schema)}.#{quote_column_name(child_index)}
+      SQL
+    end
+
+    if !_index_exists?(schema, parent_index, check_validity: true)
+      raise PgHaMigrations::InvalidMigrationError, "Parent index #{parent_index} is invalid"
```

Comment:
This should never happen, right?

Reply:
With doing no research whatsoever, my gut tells me that the implementation for partitioned indexes in postgres is fairly complicated and there are probably some conditions we're not thinking about that would cause the index to be invalid. In theory, this should never happen, but I think it's best to err on the side of caution and double check that everything worked as expected. We absolutely do not want a silent failure where this method succeeds, the index is invalid, and the user has no idea.

Reply:
Yeah, that's fine, but I think we should make that clear in the message then. E.g., something like "Unexpected state: <what you already have>"